sklearn/neural_network/multilayer_perceptron.py

```diff
@@ -355,7 +355,7 @@ def _fit(self, X, y, incremental=False):
         activations = [X]
         activations.extend(np.empty((batch_size, n_fan_out))
                            for n_fan_out in layer_units[1:])
-        deltas = [np.empty_like(a_layer) for a_layer in activations]
+        deltas = [np.empty_like(a_layer) for a_layer in activations[1:]]
```

Comment:
If deltas[-1] is the one that is never used, why are you doing `activations[1:]` rather than `activations[:-1]`? I don't know this code well so sorry if this is a stupid question.

Reply:
Because activations[0] is input(X), which can be regard as const. Only  coefficient matrix need to be changed during  iteration(need delta). Use activations[1:] can keep sure that the shape of deltas[i] won’t be changed.
More specifically, delta is changed at line238,246.Like this “deltas[last] = activations[-1] - y”，I think the shape of deltas[last] should be equal to activations[-1]. I’m also new to this field.

Reply:
Still that would shift the shape of `delta[i]` by one index. If the previous version was incorrect I'm wondering why it didn't fail with a shape mismatch error, e.g. in `_backprop` computation.

Reply:
The way this is implemented in `_backprop`, this can just be:

```py
deltas = [None] * (len(activations) - 1)
```

Reply:
@thomasjpfan was that resolved in another issue? Can't find it right now.

Reply:
> @thomasjpfan was that resolved in another issue? Can't find it right now.

        It had been changed like this:
        "activations = [X] + [None] * (len(layer_units) - 1)
        deltas = [None] * (len(activations) - 1)"
        I think I can close this pull request now.But still confused how arrays stored in memory.

Reply:
Feel free to ask a more specific question. Closed.

Reply:
Resolved in https://github.com/scikit-learn/scikit-learn/pull/14536